src/Return.jl

```diff
@@ -0,0 +1,113 @@
+"""
+    @match_fail
+
+This statement permits early-exit from the value of a @match case.
+The programmer may write the value as a `begin ... end` and then,
+within the value, the programmer may write
+
+    @match_fail
+
+to cause the case to terminate as if its pattern had failed.
+This permits cases to perform some computation before deciding if the
+rule "*really*" matched.
+"""
+macro match_fail()
+    # These are rewritten during expansion of the `@match` macro,
+    # so the actual macro should not be used directly.
+    error("$(__source__.file):$(__source__.line): @match_fail may only be used within the value of a @match case.")
+end
+
+"""
+    @match_return value
+
+This statement permits early-exit from the value of a @match case.
+The programmer may write the value as a `begin ... end` and then,
+within the value, the programmer may write
+
+    @match_return value
+
+to terminate the value expression **early** with success, with the
+given value.
+"""
+macro match_return(x)
+    # These are rewritten during expansion of the `@match` macro,
+    # so the actual macro should not be used.
+    error("$(__source__.file):$(__source__.line): @match_return may only be used within the value of a @match case.")
+end
+
+#
+# We implement @match_fail and @match_return as follows:
+#
+# Given a case (part of a @match)
+#
+#    pattern => value
+#
+# in which the value part contains a use of one of these macros, we create
+# two synthetic names: one for a `label`, and one for an intermediate `temp`.
+# Then we rewrite `value` into `new_value` by replacing every occurrence of
+#
+#    @match_return value
+#
+# with
+#
+#    begin
+#        $temp = $value
+#        @goto $label
+#    end
+#
+# and every occurrence of
+#
+#    @match_fail
+#
+# With
+#
+#    @match_return $MatchFailure
+#
+# And then we replace the whole `pattern => value` with
+#
+#    pattern where begin
+#        $temp = $value'
+#        @label $label
+#        $tmp !== $MatchFailure
+#        end => $temp
+#
+# Note that we are using the type `MatchFailure` as a sentinel value to indicate that the
+# match failed.  Therefore, don't use the @match_fail and @match_return macros for cases
+# in which `MatchFailure` is a possible result.
+#
+function adjust_case_for_return_macro(__module__, pattern, result)
+    value = gensym("value")
+    label = gensym("label")
+    found_early_exit::Bool = false
+    function adjust_top(p)
+        p isa Expr && p.head == :macrocall || return p
+        if length(p.args) == 3 &&
+            (p.args[1] == :var"@match_return"  || p.args[1] == Expr(:., Symbol(string(@__MODULE__)), QuoteNode(:var"@match_return")))
+            # :(@match_return e) -> :($value = $e; @goto $label)
+            found_early_exit = true
+            Expr(:block, p.args[2], :($value = $(p.args[3])), :(@goto $label))
```

Comment:
I'd add explicit `return` statements to these as the `if` is quite long and it's not obvious (at least to me) that it's the terminal expression in the function